### PR TITLE
kNN exact matching classifications

### LINKED ISSUE #2168: kNN classifier should allow for exact matching
Exact matching would be a useful parameter for the [kNN classifier](https://github.com/numenta/nupic/blob/master/nupic/algorithms/KNNClassifier.py); i.e. classifications where the distance is zero.

### PR DESCRIPTION
**Pull request by BoltzmannBrain**
Fixes #2168.
@subutai please review.

### CODE REVIEW

**Comment by subutai** [comment]
Could you add a test case for this option? The KNN tests are in `$NUPIC/tests/integration/nupic/algorithms/knn_classifier_test`

File: nupic/algorithms/KNNClassifier.py [hunk 1]
```diff
@@ -565,27 +569,35 @@ def infer(self, inputPattern, computeScores=True,
...
+        sorted = dist.argsort()
+        for j in sorted[:min(self.k, validVectorCount)]:
+          inferenceResult[self._categoryList[j]] += 1.0
+
+      # Prepare inference results.
+      if inferenceResult.any():
+        winner = inferenceResult.argmax()
+        inferenceResult /= inferenceResult.sum()
+      else:
+        winner = None
```
Comment (subutai) [inline nupic/algorithms/KNNClassifier.py]: You are changing the return semantics. Please update the comment for infer() with the meaning of None.  Also, add a test case that covers this scenario.

**Comment by subutai** [comment]
:+1: Other than the two comments.

**Comment by BoltzmannBrain** [comment]
@subutai please review the tests. For now I added tests for exact matching within the current [classifier_test.py](https://github.com/numenta/nupic/blob/master/tests/integration/nupic/algorithms/knn_classifier_test/classifier_test.py), but I think this file and [categories_test.py](https://github.com/numenta/nupic/blob/master/tests/integration/nupic/algorithms/knn_classifier_test/categories_test.py) should be redone to follow standard testing framework; as is it's difficult to follow what's going on. If you agree I'll go ahead and create an issue, although fixing it won't be a top priority of mine right now.

**Comment by BoltzmannBrain** [comment]
@subutai please review. To test exact matching, the classifier is trained on `patternDict` patterns and then tested on `testDict` patterns, where corresponding patterns are different for 2% of their entries. That is, for this test where each pattern is 100 values, the test pattern is identical to the training pattern for 98 of those values. Then the expected classifications for the `testDict` are all None.

**Comment by subutai** [comment]
:+1:

**Pull request merged**